File: zjctools/src/main/java/com/android/zjctools/widget/ZItemDecoration.java
package com.android.zjctools.widget;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.support.annotation.ColorInt;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

/**
 * RecyclerView列表分割线，用于LinearLayoutManager布局
 */
public class ZItemDecoration extends RecyclerView.ItemDecoration {

    private static final int HORIZONTAL_LIST = LinearLayoutManager.HORIZONTAL;
    private static final int VERTICAL_LIST = LinearLayoutManager.VERTICAL;

    private int mSpace = 1;     //间隔
    private Rect mRect = new Rect(0, 0, 0, 0);
    private Paint mPaint = new Paint();

    private int mOrientation;

    /**
     * 生成竖向排列的Decoration
     *
     * @param color  分割线颜色 例如：Color.parseString("#ff3300")
     * @param height 分割线高度（单位为PX）
     */
    public static ZItemDecoration createVertical(Context context, @ColorInt int color, int height) {
        return new ZItemDecoration(context, VERTICAL_LIST, color, height);
    }

    /**
     * 生成横向排列的Decoration
     *
     * @param color 分割线颜色 例如：Color.parseString("#ff3300")
     * @param width 分割线宽度（单位为PX）
     */
    public static ZItemDecoration createHorizontal(Context context, @ColorInt int color, int width) {
        return new ZItemDecoration(context, HORIZONTAL_LIST, color, width);
    }

    private ZItemDecoration(Context context, int orientation, @ColorInt int color, int space) {
        mOrientation = orientation;
        if (space > 0) {
            mSpace = space;
        }
        mPaint.setColor(color);
    }

    @Override
    public void onDraw(Canvas c, RecyclerView parent, RecyclerView.State state) {
        super.onDraw(c, parent, state);
        if (mOrientation == VERTICAL_LIST) {
            drawVertical(c, parent);
        } else {
            drawHorizontal(c, parent);
        }
    }

    public void drawVertical(Canvas c, RecyclerView parent) {
        final int left = parent.getPaddingLeft();
        final int right = parent.getWidth() - parent.getPaddingRight();

        final int childCount = parent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            final View child = parent.getChildAt(i);
            final RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child
                    .getLayoutParams();
            final int top = child.getBottom() + params.bottomMargin;
            final int bottom = top + mSpace;
            mRect.set(left, top, right, bottom);
            c.drawRect(mRect, mPaint);
        }
    }

    public void drawHorizontal(Canvas c, RecyclerView parent) {
        final int top = parent.getPaddingTop();
        final int bottom = parent.getHeight() - parent.getPaddingBottom();

        final int childCount = parent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            final View child = parent.getChildAt(i);
            final RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child
                    .getLayoutParams();
            final int left = child.getRight() + params.rightMargin;
            final int right = left + mSpace;
            mRect.set(left, top, right, bottom);
            c.drawRect(mRect, mPaint);
        }
    }

    @Override
    public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state) {
        super.getItemOffsets(outRect, view, parent, state);
        if (mOrientation == VERTICAL_LIST) {
            outRect.set(0, 0, 0, mSpace);
        } else {
            outRect.set(0, 0, mSpace, 0);
        }
    }
}

File: zjctools/src/main/java/com/android/zjctools/widget/ZItemDecoration.kt
package com.android.zjctools.widget

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.view.View
import androidx.annotation.ColorInt
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration

/**
 * RecyclerView列表分割线，用于LinearLayoutManager布局
 */
class ZItemDecoration(context :Context ,  orientation:Int, @ColorInt  color:Int,  space:Int)  : ItemDecoration() {


    var mOrientation = orientation
    var mSpace=space
    private val mRect = Rect(0, 0, 0, 0)
    private val mPaint = Paint()
    init {
        mPaint.color = color
    }
    companion object {
        private const val HORIZONTAL_LIST = LinearLayoutManager.HORIZONTAL
        private const val VERTICAL_LIST = LinearLayoutManager.VERTICAL

        /**
         * 生成竖向排列的Decoration
         *
         * @param color  分割线颜色 例如：Color.parseString("#ff3300")
         * @param height 分割线高度（单位为PX）
         */
        fun createVertical(context: Context, @ColorInt color: Int, height: Int): ZItemDecoration {
            return ZItemDecoration(context, VERTICAL_LIST, color, height)
        }

        /**
         * 生成横向排列的Decoration
         *
         * @param color 分割线颜色 例如：Color.parseString("#ff3300")
         * @param width 分割线宽度（单位为PX）
         */
        fun createHorizontal(context: Context, @ColorInt color: Int, width: Int): ZItemDecoration {
            return ZItemDecoration(context, HORIZONTAL_LIST, color, width)
        }
    }


    override fun onDraw(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        super.onDraw(c, parent, state)
        if (mOrientation == VERTICAL_LIST) {
            drawVertical(c, parent)
        } else {
            drawHorizontal(c, parent)
        }
    }

    fun drawVertical(c: Canvas, parent: RecyclerView) {
        val left = parent.paddingLeft
        val right = parent.width - parent.paddingRight
        val childCount = parent.childCount
        for (i in 0 until childCount) {
            val child = parent.getChildAt(i)
            val params = child
                .layoutParams as RecyclerView.LayoutParams
            val top = child.bottom + params.bottomMargin
            val bottom = top + mSpace
            mRect[left, top, right] = bottom
            c.drawRect(mRect, mPaint)
        }
    }

    fun drawHorizontal(c: Canvas, parent: RecyclerView) {
        val top = parent.paddingTop
        val bottom = parent.height - parent.paddingBottom
        val childCount = parent.childCount
        for (i in 0 until childCount) {
            val child = parent.getChildAt(i)
            val params = child
                .layoutParams as RecyclerView.LayoutParams
            val left = child.right + params.rightMargin
            val right = left + mSpace
            mRect[left, top, right] = bottom
            c.drawRect(mRect, mPaint)
        }
    }

    override fun getItemOffsets(
        outRect: Rect,
        view: View,
        parent: RecyclerView,
        state: RecyclerView.State) {
        super.getItemOffsets(outRect, view, parent, state)
        if (mOrientation == VERTICAL_LIST) {
            outRect[0, 0, 0] = mSpace
        } else {
            outRect[0, 0, mSpace] = 0
        }
    }




}